crates/host/src/wasmbus/mod.rs

```diff
@@ -0,0 +1,642 @@
+            Poll::Ready(None) => {}
+            Poll::Pending => pending = true,
+        }
+        match Pin::new(&mut self.inventory).poll_next(cx) {
+            Poll::Ready(Some(msg)) => return Poll::Ready(Some(msg)),
+            Poll::Ready(None) => {}
+            Poll::Pending => pending = true,
+        }
+        match Pin::new(&mut self.auction).poll_next(cx) {
+            Poll::Ready(Some(msg)) => return Poll::Ready(Some(msg)),
+            Poll::Ready(None) => {}
+            Poll::Pending => pending = true,
+        }
+        match Pin::new(&mut self.pings).poll_next(cx) {
+            Poll::Ready(Some(msg)) => return Poll::Ready(Some(msg)),
+            Poll::Ready(None) => {}
+            Poll::Pending => pending = true,
+        }
+        if pending {
+            Poll::Pending
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+struct AsyncBytesMut(Arc<std::sync::Mutex<BytesMut>>);
+
+impl AsyncWrite for AsyncBytesMut {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        Poll::Ready({
+            self.0
+                .lock()
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))?
+                .put_slice(buf);
+            Ok(buf.len())
+        })
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), std::io::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl TryFrom<AsyncBytesMut> for Vec<u8> {
+    type Error = anyhow::Error;
+
+    fn try_from(buf: AsyncBytesMut) -> Result<Self, Self::Error> {
+        buf.0
+            .lock()
+            .map(|buf| buf.clone().into())
+            .map_err(|e| anyhow!(e.to_string()).context("failed to lock"))
+    }
+}
+
+impl Queue {
+    #[instrument]
+    async fn new(
+        nats: &async_nats::Client,
+        cluster_key: &KeyPair,
+        host_key: &KeyPair,
+    ) -> anyhow::Result<Self> {
+        let (registries, pings, links, queries, auction, commands, inventory) = try_join!(
+            nats.subscribe("wasmbus.ctl.default.registries.put".into()),
+            nats.subscribe("wasmbus.ctl.default.ping.hosts".into()),
+            nats.subscribe("wasmbus.ctl.default.linkdefs.*".into()),
+            nats.subscribe("wasmbus.ctl.default.get.*".into()),
+            nats.subscribe("wasmbus.ctl.default.auction.>".into()),
+            nats.subscribe(format!(
+                "wasmbus.ctl.default.cmd.{}.*",
+                host_key.public_key()
+            )),
+            nats.subscribe(format!(
+                "wasmbus.ctl.default.get.{}.inv",
+                host_key.public_key()
+            )),
+        )
```

Comment:
The `default` in each of these should eventually be a parameter that's passed in and defined at runtime, to match the current functionality. It's the lattice prefix, but as the PR text mentions this isn't the full functionality so I'm just calling attention to it 👍🏻